src/main/kotlin/io/perfometer/runner/ThreadPoolScenarioRunner.kt

```diff
@@ -1,62 +1,95 @@
+    private val parallelJobsExecutor: ExecutorService = Executors.newCachedThreadPool()
+
     override fun runUsers(
         userCount: Int,
         duration: Duration,
         action: suspend () -> Unit,
     ): ScenarioSummary {
-        val scenarioExecutor = Executors.newFixedThreadPool(userCount)
+        return Executors.newFixedThreadPool(userCount).let { executor ->
+            runUsersInternal(userCount, executor, action).let { usersFeature ->
+                timeoutExecutors(duration, executor)
+                usersFeature.join()
+                statistics.finish()
+            }
+        }
+    }
+
+    override suspend fun runStep(step: HttpStep) {
+        when (step) {
+            is RequestStep -> executeHttp(step)
+            is PauseStep -> pauseFor(step.duration)
+            is ParallelStep -> runParallel(step)
+        }
+    }
 
-        val future = CompletableFuture.allOf(
+    override suspend fun runStepAsync(step: HttpStep) {
+        parallelJobs.add(
+            CompletableFuture.runAsync(
+                { runBlocking { runStep(step) } }, parallelJobsExecutor
+            )
+        )
+    }
+
+    private fun runUsersInternal(
+        userCount: Int,
+        scenarioExecutor: ExecutorService,
+        action: suspend () -> Unit,
+    ): CompletableFuture<Void> {
+        return CompletableFuture.allOf(
             *(0 until userCount)
                 .map { CompletableFuture.runAsync({ runAction(action) }, scenarioExecutor) }
                 .toTypedArray())
+    }
 
+    private fun timeoutExecutors(
+        duration: Duration,
+        scenarioExecutor: ExecutorService,
+    ) {
         Executors.newSingleThreadScheduledExecutor().schedule({
             scenarioExecutor.shutdownNow()
+            parallelJobsExecutor.shutdownNow()
         }, duration.toNanos(), TimeUnit.NANOSECONDS)
-
-        future.join()
-        return statistics.finish()
     }
 
-    private fun runAction(action: suspend () -> Unit) {
+    private fun runParallel(step: ParallelStep) {
         runBlocking {
-            while (!Thread.currentThread().isInterrupted) {
-                try {
-                    action()
-                } catch (e: InterruptedException) {
-                    Thread.currentThread().interrupt()
-                }
-            }
+            step.action()
         }
+        CompletableFuture.allOf(*parallelJobs.toTypedArray()).join()
     }
 
-    override suspend fun runStep(step: HttpStep) {
-        when (step) {
-            is RequestStep -> executeHttp(step)
-            is PauseStep -> pauseFor(step.duration)
+    private fun runAction(action: suspend () -> Unit) = decorateInterruptable {
+        runBlocking {
+            while (!Thread.currentThread().isInterrupted) {
+                decorateSuspendingInterruptable(action)
+            }
         }
     }
 
-    private fun pauseFor(duration: Duration) {
+    private fun pauseFor(duration: Duration) = decorateInterruptable {
```

Comment:
Why do we need this decoration here?

Reply:
In case where parallel block runs longer than the scenario is planned to, and sleep is interrupted, `InteruptedException` is thrown from `Thread.sleep()`. 

Pausing indeed doesn't make sense `parallel` blocks, so I'll remove this anyways.